Update RNA-seq gene ID conversion example (#218)
AlexsLemonade/refinebio-examples: 03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

---
title: "Obtaining Annotation for Ensembl IDs - RNA-seq"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of this analysis

> The purpose of this notebook is to provide an example of mapping gene IDs for RNA-seq data obtained from refine.bio using `AnnotationDbi` packages.

Reviewer: Probably should have a citation here.

Experiment accession IDs usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `SRP056136` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis `.Rmd` you downloaded  
- A folder called "data" which contains:
  - The `SRP056136` folder which contains:
    - The gene expression  
    - The metadata TSV  
- A folder for `plots` (currently empty)  
- A folder for `results` (currently empty)  

Your example analysis folder should now look something like this (except with respective experiment accession ID and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Define the file path to the data directory

> data_dir <- file.path("data", "SRP056136")

Reviewer: I think there were spaces in between these lines?

library(org.Dr.eg.db)
# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP056136.tsv" # Replace with the name of your metadata file
))
# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP056136.tsv" # Replace with the name of your data file
)) %>%
  # Tuck away the Gene ID column as rownames
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)

# Bring back the "Gene" column in preparation for mapping
df <- df %>%
  tibble::rownames_to_column("Gene")
```

> ## Map Ensembl IDs to Uniprot IDs

Reviewer: IDK if mapping to protein ids is a great idea. In general RNA and protein are pretty poorly correlated, so we may not want to encourage this science wise, though it is a thing you can do. 

What about RefSeq ids or another RNA-based id? 

Do the RefSeq IDs only have `NM_` accessions? and not `XM_` or etc?

Reviewer 2: Let's do Entrez IDs - looks like #212 includes a symbols example. I expect these are going to be the most common use cases.

Author: I can try RefSeq again, but it was so memory intensive that it didn't complete locally on my machine. I can give it another try but I am not sure that's the way to go here.

Reviewer: Perfect!

  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/SRP056136/rna-seq-of-control-and-foxo3b-mo-zebrafish-cells).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-email.png" width=500>  

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

> For this example analysis, we will use this [RNA-seq zebrafish dataset](https://www.refine.bio/experiments/SRP056136/rna-seq-of-control-and-foxo3b-mo-zebrafish-cells).

Reviewer: I'm wondering if we want to use a bigger dataset. n = 4 is pretty small, although it doesn't really matter since we won't be using the data itself for anything. 

May want to add a bit more about the dataset either way.

Author: I had the same thoughts and ended up leaving the dataset as is. Good catch, more description is definitely needed here either way. 

Not sure if this means we should dig into finding a new dataset however.

Reviewer: I think switching to a dataset users can use for more than just this might be good. This is the time to make these changes I suppose.

Author: Gotcha, I'll look for one and post what I found on the issue relevant to this PR.

```{r}
# Attach the library
library(org.Dr.eg.db)
# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP040561.tsv" # Replace with the name of your metadata file
))
# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP040561.tsv" # Replace with the name of your data file
)) %>%
  # Tuck away the Gene ID column as rownames
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)

# Bring back the "Gene" column in preparation for mapping

> df <- df %>%

Author: I left it as is because thinks looked a little more complex with the negative indexing (I agree, either way isn't perfect but at least this way we can stick with tidyverse).

> df <- df %>%

Reviewer: I'm fine with that, and if the next reviewer disagrees, then we can revisit it.

Let's get a summary of the values returned in the `Entrez` column of our mapped data frame.

```{r}
# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
summary(mapped_df$Entrez)
```

There are 9088 NA's in our data frame, which means that 9088 out of the 31882 Ensembl IDs did not map to Entrez IDs.
This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.

Now let's check to see how many genes we have that were mapped to multiple IDs.

```{r}
multi_mapped_df <- mapped_df %>%
  # Let's group by the Ensembl ID's in the `Ensembl` column
  dplyr::group_by(Ensembl) %>%
  # Create a new variable containing the number of Entrez IDs mapped to each ID
  dplyr::mutate(entrez_id_count = dplyr::n()) %>%
  # Arrange by the genes with the highest number of Entrez IDs mapped
  dplyr::arrange(desc(entrez_id_count)) %>%
  # Join the remaining data from `df` using the Ensembl IDs

>   dplyr::inner_join(df, by = c("Ensembl" = "Gene"))

Reviewer: If we go with either of the options I mentioned, we could drop the inner_join here OR we could keep it so its abundantly clear that multi-mappings all end up having the same values.

Author: I agree with both points here, however, I chose to drop it in the last commit because I do not think it is completely necessary here as they are able to see that multi-mappings have the same values in the `all_mapped_df` object (if we choose to drop that object, we may want to revisit this comment/thought).

# Let's look at the first 6 rows of our `multi_mapped` object
head(multi_mapped_df)
```

Looks like we have some cases where 7 Entrez IDs mapped to a single Ensembl ID.
We have a total of 542 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument og `mapIds()`, like we did in the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).

However, we may also want to keep the multi-mapped information for our downstream analyses so let's go ahead and write our mapped results and data to file!

## Write mapped results to file

```{r}
# Write mapped and annotated data frame to output file

> readr::write_tsv(multi_mapped_df, file.path(

Reviewer: I don't really like the idea of keeping so many rows of redundant data, for fear that a user might think its appropriate to use this dataset (with duplicate rows of data) elsewhere. Keeping duplicate rows of values feels not ideal since we didn't actually measure each of these Entrez genes separately. However, I'm also not sure of any rationale to decide on Entrez ID over another. 

I have two ideas, neither of which I think are that great: 

1) We tell the user, we don't have a particular reason to choose one over the other, so let's just go with the first Entrez ID that comes up, then we re-run with `multiVals = "first"`, save to TSV.  

2) We say, we're not really sure which Entrez ID is relevant, but we'd like to store this information for future use. Then we can collapse all the Entrez IDs into one column where we store them all. We could do this by.   
```
df %>% 
  dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = "-")) %>% 
  dplyr::select(entrez_ids, dplyr::everything())
```

Option 1 would be easier for the user to use in the future, but if they can't find a Entrez ID they are looking for, they night have issues figuring that out but Option 2 allows them to keep all the information, should they need to access it. (However, I am concerned a concatenated column might not be easy for our users to use ether).

Author: I agree that Option 1 would be easier for downstream use, while Option 2 allows them to keep all the information which many people would probably desire (even if just to save in a TSV or CSV format).

I incorporated both ideas in the last commit but end up saving Option 1 to the TSV file. I tried to keep it as short as possible but let me know if you still think we should just choose one! (Then we will probably revisit your [comment on keeping/dropping the `inner_join` of the expression values](https://github.com/AlexsLemonade/refinebio-examples/pull/218#discussion_r490950035)).

Reviewer: We should probably choose one so users aren't confused on why we are making two data.frames and only saving one. I'm not sure which one though.

Reviewer 3: I would probably do a hybrid: a column with the first (or best?) Entrez ID and one with all, separated by semicolons (Hyphens will be confusing, I think). I do think there should be some discussion of how you *might* pick one, with first as one option, but some other potential reasons discussed. Maybe you have a list of Entrez IDs you are interested in? Maybe you want the longest version of the gene? Maybe you want the oldest Entrez ID (which is probably first). I don't think we need to make a recommendation, but I do think there should be some acknowledgement of the issue.

```{r}
# Map Ensembl IDs to their associated Entrez IDs
mapped_list <- mapIds(
  org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
  keys = df$Gene,
  column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
  multiVals = "list"
)
```

## Explore gene ID conversion

Now, let's take a look at our mapped object to see how the mapping went.

```{r}
# Let's use the `head()` function to take a preview at our mapped list
head(mapped_list)
```

It looks like we have Entrez IDs that were successfully mapped to the Ensembl IDs we provided.
However, the data is now in a `list` object, making it a little more difficult to explore.
We are going to turn our list object into a data frame object in the next chunk.

```{r}
# Let's make our object a bit more manageable for exploration by turning it into a data frame using the `reshape2::melt()` function

> mapped_df <- mapped_list %>%
>   reshape2::melt(value.name = "Entrez") %>%
>   dplyr::rename("Ensembl" = "L1") # Here we are renaming the column holding the Ensembl IDs as good practice

Reviewer 3: We can do this in a more "modern" way without having to install `reshape2`
```suggestion
mapped_df <- mapped_list %>%
  tibble::enframe(name = "Ensembl", value = "Entrez") %>%
  tidyr::unnest() # enframe makes a `list` column, so here we are converting that to simpler format
```

```

Now let's take a peek at our data frame.

```{r}
head(mapped_df)
```

We can see that our data frame has a new column `Entrez`.
Let's get a summary of the values returned in the `Entrez` column of our mapped data frame.

```{r}
# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
summary(as.factor(mapped_df$Entrez)) # We need to use the `as.factor()` function here in order to get the count of each unique value returned
```

There are 9102 NA's in our data frame, which means that 9102 out of the 31882 Ensembl IDs did not map to Entrez IDs.
This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.

Now let's check to see how many genes we have that were mapped to multiple IDs.

```{r}

> multi_mapped_df <- mapped_df %>%
>   # Let's group by the Ensembl ID's in the `Ensembl` column
>   dplyr::group_by(Ensembl) %>%
>   # Create a new variable containing the number of Entrez IDs mapped to each ID
>   dplyr::mutate(entrez_id_count = dplyr::n()) %>%
>   # Arrange by the genes with the highest number of Entrez IDs mapped
>   dplyr::arrange(desc(entrez_id_count))

Reviewer 3: The resulting data table here is a bit confusing, as it still has both the Ensembl ID and the Entrez ID, so there are multiple rows for each Ensembl ID. The smallest change is below, using `summarize` instead of `mutate`:

```suggestion
multi_mapped_df <- mapped_df %>%
  # Let's group by the Ensembl ID's in the `Ensembl` column
  dplyr::group_by(Ensembl) %>%
  # Create a new variable containing the number of Entrez IDs mapped to each ID
  dplyr::summarize(entrez_id_count = dplyr::n()) %>%
  # Arrange by the genes with the highest number of Entrez IDs mapped
  dplyr::arrange(desc(entrez_id_count))
```

But if we want to be fancy, we can make it simpler with `count`
```
multi_mapped_df <- mapped_df %>%
  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
  dplyr::count(Ensembl, name = "entrez_id_count") %>%
  dplyr::arrange(desc(entrez_id_count))
```

```{r}
# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
summary(as.factor(mapped_df$Entrez)) # We need to use the `as.factor()` function here in order to get the count of each unique value returned
```

There are 9102 NA's in our data frame, which means that 9102 out of the 31882 Ensembl IDs did not map to Entrez IDs.
This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.

Now let's check to see how many genes we have that were mapped to multiple IDs.

```{r}
multi_mapped_df <- mapped_df %>%
  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
  dplyr::count(Ensembl, name = "entrez_id_count") %>%
  # Arrange by the genes with the highest number of Entrez IDs mapped
  dplyr::arrange(desc(entrez_id_count))

# Let's look at the first 6 rows of our `multi_mapped` object
head(multi_mapped_df)
```

> Looks like we have some cases where 15 Entrez IDs mapped to a single Ensembl ID.

Reviewer 3: Keeping the exclamation point, but changing direction to match the way we are going...
```suggestion
Looks like we have one case where one Ensembl id mapped to 15 Entrez IDs!
```

We have a total of 235 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.
If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).

### Store all mapped values

We are not sure which Entrez ID is relevant as yet, but we could store all of the mapped information for future use.
In the next chunk, we show how we can collapse all the Entrez IDs into one column where we store them all for each unique Ensembl ID.

```{r}
final_mapped_df <- df %>%
  # Collapse the mapped values in `mapped_list` into one column named `entrez_ids`

>   dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = ";")) %>%

Reviewer: ```suggestion
  dplyr::mutate(all_entrez_ids = sapply(mapped_list, paste, collapse = ";")) %>%
```
Since below you named the next column, `first_entrez_ids`, you can name this one in parallel. (If you do this you'll need to change the next couple lines too.

Reviewer: Including the comments!

>   dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = ";")) %>%

Reviewer 3: Related to my fear of order changes, I wouldn't trust this to return a list in the same order as `df`, so I would probably do this separately and join them.

Since you already have this in a data frame, from the step before, you could do this with `dplyr::summarize` pretty neatly: 
```
collapsed_mapped_df <- mapped_df %>%
  dplyr::group_by(Ensembl) %>%
  dplyr::summarize(entrez_ids = paste(Entrez, collapse = ";"))
```

```

Let's take a look at our new collapsed `entrez_ids` column in the `final_mapped_df` object.

```{r}
final_mapped_df %>%
  # Filter `final_mapped_df` to include only the rows where `entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
  dplyr::filter(stringr::str_detect(entrez_ids, ";")) %>%
  # We only need a preview here
  head()
```

You may have a list of Entrez IDs you are interested in, in which case, the above format may work best for you.
In a different study, you may want the oldest Entrez ID (which is probably first), in which case, you can create a column that stores just the first mapped Entrez ID that comes up for each Ensembl ID.
We will do this in the next section by re-running the `mapIds()` function with `multiVals = "first"`.

### Map Ensembl IDs to gene symbols -- keeping only the first mapped value

Let's re-run `mapIds()`, this time using `multiVals = "first"` and add this information to our `final_mapped_df` object.

```{r}
final_mapped_df <- data.frame(
  "first_mapped_entrez_id" = mapIds(
    org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data

>     keys = df$Gene,

Reviewer: ```suggestion
    keys = final_mapped_df$Gene,
```

    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID

>   )

Reviewer: ```suggestion
  ), final_mapped_df
```

Then no `inner_join` is needed. (I think this should be fine). Unless we find it too confusing for users to follow? (II personally find it easier to follow than a join.

Reviewer 3: I think this would result in a duplicated column if you leave the `rownames_to_column` step... but I like joins. They are a good thing to learn and a good habit because I have been burned too many times by functions that don't preserve order (`mapIds` seems to, but I still don't trust it).

) %>%
  # Make an `Ensembl` column to store the rownames
  tibble::rownames_to_column("Ensembl") %>%
  # Join the remaining data from `df` using the Ensembl IDs

>   dplyr::inner_join(final_mapped_df, by = c("Ensembl" = "Gene"))

Reviewer: I haven't run this locally, but I don't think you need to `inner_join()` this? I think you will already have the data from before?

Reviewer 3: It is always safer to use a join in case order has changed, so I would leave it 😄

Reviewer: The order doesn't change based on my suggestion because you'd directly provide final_mapped_df$Ensembl (or whatever the correct column name is) to be mapped.

Reviewer 3: `final_mapped_df$Ensembl` is not guaranteed  to be in the same order as `df$Gene`. In fact, in this case `final_mapped_df` was sorted by Entrez_id on line 310, so I think it *won't* be in the same order. The fact that this kind of thing is hard to keep track of is exactly why I favor joins.

Reviewer 3: Oh, I see what you are saying... I missed that change. But I stand by my general point, because I still don't trust `mapIds` not to rearrange things.

>   tibble::rownames_to_column("Ensembl") %>%
>   # Join the remaining data from `df` using the Ensembl IDs
>   dplyr::inner_join(final_mapped_df, by = c("Ensembl" = "Gene"))

Reviewer 3: Just for consistency, since the other columns will have `id` in the name.
```suggestion
  tibble::rownames_to_column("Ensembl_id") %>%
  # Join the remaining data from `df` using the Ensembl IDs
  dplyr::inner_join(final_mapped_df, by = c("Ensembl_id" = "Gene"))
```

Reviewer 3: I maybe don't like this after seeing how the data comes out of mapIds. Ignore this and stick with just plain `Ensembl`

```

Our `final_mapped_df` object now has a column named `first_mapped_entrez_id` that contains only the first mapped Entrez ID, in addition to the `entrez_ids` column that contains all mapped Entrez IDs per Ensembl ID.

```{r}
final_mapped_df %>%

>   dplyr::filter(stringr::str_detect(entrez_ids, ";")) %>%

Reviewer: Is this filter just so you can see the preview of multiples? We may wanna put a comment to explain that.

if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/SRP040561/gene-expression-analysis-of-hair-cell-regeneration-in-the-zebrafish-lateral-line).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-email.png" width=500>  

If you want to use this dataset with other examples you may want to download it non-quantile normalized.

> See the [RNA-seq header section](https://github.com/AlexsLemonade/refinebio-examples/tree/master/03-rnaseq/00-intro-to-rnaseq.html#14About_quantile_normalization) about quantile normalization.

Reviewer 3: Fix link?
```suggestion
See the [RNA-seq header section](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html#about-quantile-normalization) about quantile normalization.
```

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

refine.bio data comes with gene level data with Ensembl IDs. 
Although this example script uses Ensembl IDs from Zebrafish, (<i>Danio rerio</i>), to obtain Entrez IDs this script can be easily converted for use with different species or annotation types <i>eg</i> protein IDs, gene ontology, accession numbers.

<b>For different species</b>, wherever the abbreviation `org.Dr.eg.db` or `Dr` is written, it must be replaced with the respective species abbreviation <i>eg</i> for<i> Homo sapiens</i> `org.Hs.eg.db` or `Hs` would be used.

> In the case of our [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html), a Mouse (<i>Mus musculus</i>) dataset is used, meaning `org.Mm.eg.db` or `Mm` would also need to be used there.

Reviewer 3: Another link fix.
```suggestion
In the case of our [microarray gene identifier annotation example notebook](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html), a Mouse (<i>Mus musculus</i>) dataset is used, meaning `org.Mm.eg.db` or `Mm` would also need to be used there.
```

A full list of the annotation R packages from Bioconductor is at this [link](https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData) [@annotation-packages].

***

<!-- Do not delete this line --> <a name="analysis" style="padding-top:56px;margin-top:-56px;">&nbsp;</a>

# Obtaining Annotation for Ensembl IDs - RNA-seq

Ensembl IDs can be used to obtain various different annotations at the gene/transcript level.

> Let's get ready to use the Ensembl IDs from our mouse dataset to obtain the associated Entrez IDs.

Author: Good catch!

```{r}
# Map Ensembl IDs to their associated Entrez IDs
mapped_list <- mapIds(
  org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
  keys = df$Gene,
  column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
  multiVals = "list"
)
```

## Explore gene ID conversion

Now, let's take a look at our mapped object to see how the mapping went.

```{r}
# Let's use the `head()` function to take a preview at our mapped list
head(mapped_list)
```

It looks like we have Entrez IDs that were successfully mapped to the Ensembl IDs we provided.
However, the data is now in a `list` object, making it a little more difficult to explore.
We are going to turn our list object into a data frame object in the next chunk.

```{r}
# Let's make our object a bit more manageable for exploration by turning it into a data frame
mapped_df <- mapped_list %>%
  tibble::enframe(name = "Ensembl", value = "Entrez") %>%

>   tidyr::unnest(cols = Entrez) # enframe makes a `list` column, so here we are converting that to simpler format

Reviewer 3: I thought this comment could use a bit more detail.
```suggestion
  # enframe makes a `list` column, so we will convert that to simpler format with `unnest()
  # This will result in one row of our data frame per list item
  tidyr::unnest(cols = Entrez) 
```

```{r}
# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
summary(as.factor(mapped_df$Entrez)) # We need to use the `as.factor()` function here in order to get the count of each unique value returned
```

There are 9102 NA's in our data frame, which means that 9102 out of the 31882 Ensembl IDs did not map to Entrez IDs.
This means if you are depending on `Entrez` IDs for your downstream analyses, you may not be able to use the 9102 unmapped genes.

Now let's check to see how many genes we have that were mapped to multiple IDs.

```{r}
multi_mapped_df <- mapped_df %>%
  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
  dplyr::count(Ensembl, name = "entrez_id_count") %>%
  # Arrange by the genes with the highest number of Entrez IDs mapped
  dplyr::arrange(desc(entrez_id_count))

# Let's look at the first 6 rows of our `multi_mapped` object
head(multi_mapped_df)
```

Looks like we have one case where one Ensembl ID mapped to 15 Entrez IDs!
We have a total of 235 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.

> With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.

Reviewer 3: Is this something we would really want to do? I feel like suggesting filtering out the multimaps is dangerous...I see that we do it elsewhere, but I am not sure I am keen on it. 

More relevant here, I think this choice probably not particularly dependent on the rate so much as what genes would be lost, so I might suggest this with more ambiguity as to why you might do it: 
```suggestion
One option is to filter out the multi-mapped IDs and keep only the 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.
```

> If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).

Reviewer 3: One more link fix...
```suggestion
If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
```

### Store all mapped values

If we are not sure which Entrez ID is most relevant to our downstream analyses, we could store all of the mapped information for future use.
In the next chunk, we show how we can collapse all the Entrez IDs into one column where we store them all for each unique Ensembl ID.

```{r}
collapsed_mapped_df <- mapped_df %>%
  # Group by Ensembl IDs
  dplyr::group_by(Ensembl) %>%
  # Collapse the mapped values in `mapped_df` into one column named `all_entrez_ids`
  dplyr::summarize(all_entrez_ids = paste(Entrez, collapse = ";"))
```

Let's take a look at our new collapsed `all_entrez_ids` column in the `collapsed_mapped_df` object.

```{r}
collapsed_mapped_df %>%
  # Filter `collapsed_mapped_df` to include only the rows where `all_entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
  dplyr::filter(stringr::str_detect(all_entrez_ids, ";")) %>%
  # We only need a preview here
  head()
```

You may have a list of Entrez IDs you are interested in, in which case, the above format may work best for you.

> In a different study, you may want the oldest Entrez ID (which is probably first), in which case, you can create a column that stores just the first mapped Entrez ID that comes up for each Ensembl ID.

Reviewer 3: Do we have confirmation that the first Entrez ID is the lowest number? i.e. is the list we get here sorted?

Author: Good question! It does not explicitly that the returned list is sorted in the `AnnotatationDbi` or `mapIDs()` docs but the lists of mapped values returned does appear to be sorted from my investigation.

  # Now let's join the rest of the expression data
  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
```

Our `final_mapped_df` object now has a column named `first_mapped_entrez_id` that contains only the first mapped Entrez ID, in addition to the `all_entrez_ids` column that contains all mapped Entrez IDs per Ensembl ID.

```{r}
final_mapped_df %>%
  # Filter `final_mapped_df` to preview the rows where `all_entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
  dplyr::filter(stringr::str_detect(all_entrez_ids, ";")) %>%
  head()
```

Now let's write our mapped results and data to file!

## Write mapped results to file

```{r}
# Write mapped and annotated data frame to output file
readr::write_tsv(final_mapped_df, file.path(
  results_dir,
  "SRP040561_Entrez_IDs.tsv" # Replace with a relevant output file name
))
```

# Resources for further learning

- Marc Carlson has prepared a nice [Introduction to Bioconductor Annotation Packages](https://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf) [@Carlson2020-vignette]

> - See our [microarray gene ID conversion notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html) as another applicable example, since the steps for this workflow do not change with technology [@gene-id-annotation-microarray].

Reviewer 3: ```suggestion
- See our [microarray gene ID conversion notebook](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html) as another applicable example, since the steps for this workflow do not change with technology [@gene-id-annotation-microarray].
```
